inference-engine/thirdparty/clDNN/src/layout_optimizer.cpp

```diff
@@ -738,12 +738,96 @@ layout layout_optimizer::get_expected_layout(layout const& current_layout,
     auto expected_data_type = current_layout.data_type;
     auto expected_format = current_layout.format;
     auto input_layout = node.get_dependency(0).get_output_layout();
+    auto output_layout = node.calc_output_layout();
 
     const float cond_denom = _total_conv > 0 ? 1.0f / static_cast<float>(_total_conv) : 1.0f;
 
-    auto output_layout = node.calc_output_layout();
+    bool is_dw = input_layout.size.feature[0] == static_cast<int>(prim->groups);
+    int ofm_per_group = output_layout.size.feature[0] / prim->groups;
+    int ifm_per_group = input_layout.size.feature[0] / prim->groups;
+    int compute_block = 32;
+    bool valid_grouped = !is_dw && prim->groups > 1 && (ofm_per_group % compute_block == 0 && ifm_per_group % compute_block == 0);
+    bool valid_int8_dw = is_dw && output_layout.size.batch[0] % 16 == 0;
+    bool non_grouped = prim->groups == 1;
+    bool is_2d = input_layout.format.spatial_num() == 2;
+    bool valid_post_ops = get_post_ops_count(node) <= 32;
+    bool use_onednn_impls = _optimization_attributes.use_onednn_impls;
+    bool i8_u8_input = input_layout.data_type == data_types::u8 || input_layout.data_type == data_types::i8;
+
+    if (use_onednn_impls && valid_post_ops) {
+        for (auto& fo : node.get_fused_primitives()) {
+            if (fo.node->is_type<eltwise>()) {
+                auto in_layout = node.get_dependency(fo.dep_start_idx).get_output_layout();
+                auto out_layout = node.get_output_layout();
+                auto in_dt = in_layout.data_type;
+                auto out_dt = out_layout.data_type;
+                if ((out_layout.count() == in_layout.count()) &&
+                    (data_type_traits::is_floating_point(in_dt) || data_type_traits::is_floating_point(out_dt)) && in_dt != out_dt) {
+                    valid_post_ops = false;
+                    break;
+                }
+            }
+        }
+    }
+
+    /* ***************************************************************************************** */
+    /* ***************************** OneDNN layout selection part ****************************** */
+    /* ***************************************************************************************** */
+
+    if (use_onednn_impls && i8_u8_input) {
+        if ((non_grouped || valid_grouped || valid_int8_dw) && valid_post_ops && is_2d) {
```

Comment:
We will fallback to cldnn if `valid_post_ops` is false, right? If so, I think `valid_post_ops` needs to be on outer condition so that we are following general  cldnn rule..

Reply:
Yes, but I'm afraid, that there can be some kind of "mixing" of formats selection, which may provide some benefits
For example, if we have unsupported post ops by OneDNN impls, we may get into second branch (`else if ((_optimization_attributes.b_fs_yx_fsv16_network &&...`) and select b_fs_yx_fsv16 or b_fs_yx_fsv32 for native impls (at the same time "native format selection part" doesn't contain such rule). **I didn't check it on real topologies**, but theoretically it may help us to reduce reorders or just select faster kernel in some cases
What do you think? Maybe we can postpone such cleaning, until we have enough level of test coverage?